django/template/loader_tags.py

```diff
@@ -178,12 +178,22 @@ def __init__(self, template, *args, **kwargs):
         super(IncludeNode, self).__init__(*args, **kwargs)
 
     def render(self, context):
+        """
+        Render the specified template and context. Cache the template object
+        in render_context to avoid reparsing and loading when used in a for
+        loop.
+        """
         try:
             template = self.template.resolve(context)
             # Does this quack like a Template?
             if not callable(getattr(template, 'render', None)):
-                # If not, we'll try get_template
-                template = context.template.engine.get_template(template)
+                # If not, we'll try our cache, and get_template
+                template_name = template
+                template = context.render_context.get(template_name)
+                if template is None:
+                    template = context.template.engine.get_template(template_name)
+                context.render_context[template_name] = template
```

Comment:
This is better. I didn't think about the problem with variables that can change values. I have one concern, though. The same ``render_context`` instance is shared across nodes in the ``Template.render()`` call, so ``template_name`` isn't a very safe key to use. The potential arises to clash with other nodes. Maybe we could adjust the approach a bit?

My thought is to create a cache dict in the ``render_context`` using ``self`` as a key. For example:

```
cache = context.render_context.setdefault(self, {})
template = cache.get(template_name)
cache[template_name] = template
```

Alternatively, you could use a unique, uncommon key rather than ``self``. The benefit here is that multiple ``IncludeNode`` instances could share the same cache. That means a template like below would only parse ``template.html`` once:

```
{% include "template.html" %}
{% include "template.html" %}
{% include "template.html" %}
```

That seems appropriate in this case.